lemur/auth/vault_jwt_auth.py

```diff
@@ -0,0 +1,244 @@
+
+    @property
+    def public_key_use(self):
+        return self._jwk_data.get("use", None)
+
+
+class PyJWKSet:
+    def __init__(self, keys):
+        self.keys_by_id = {}
+
+        if not keys or not isinstance(keys, list):
+            raise PyJWKSetError("Invalid JWK Set value")
+
+        if len(keys) == 0:
+            raise PyJWKSetError("The JWK Set did not contain any keys")
+
+        for key in keys:
+            k = PyJWK(key)
+            self.keys_by_id[k.key_id] = k
+
+    @staticmethod
+    def from_dict(obj):
+        keys = obj.get("keys", [])
+        return PyJWKSet(keys)
+
+
+class TimeoutRetryingHTTPAdapter(HTTPAdapter):
+    def __init__(self, timeout=None, retry_policy=None, *args, **kwargs):
+        timeout_value = timeout
+        if timeout_value is None:
+            timeout_value = _DEFAULT_TIMEOUT
+        elif not isinstance(timeout_value, tuple):
+            # a request_timeout only, use sane default connect_timeout
+            timeout_value = (_DEFAULT_TIMEOUT[0], timeout_value)
+        self.timeout = timeout_value
+
+        self.max_retries = _DEFAULT_RETRY_POLICY
+        if isinstance(retry_policy, Retry):
+            self.max_retries = retry_policy
+        super(TimeoutRetryingHTTPAdapter, self).__init__(*args, **kwargs)
+
+    def send(self, request, **kwargs):
+        timeout = kwargs.get("timeout")
+        if timeout is None:
+            kwargs["timeout"] = self.timeout
+        elif not isinstance(timeout, tuple):
+            # a request_timeout only, use sane default connect_timeout
+            kwargs["timeout"] = (_DEFAULT_TIMEOUT[0], timeout)
+        return super(TimeoutRetryingHTTPAdapter, self).send(request, **kwargs)
+
+
+class JWTAuthenticator(object):
+    _TTL_MARGIN = 300  # 5 minutes in seconds
+
+    _instances = {}
+
+    @classmethod
+    def instance(
+        cls,
+        name,
+        audience,
+        additional_issuers=None,
+        datacenter=None,
+        timeout=None,
+    ):
+        if name not in cls._instances:
+            cls._instances[name] = JWTAuthenticator(
+                audience=audience,
+                additional_issuers=additional_issuers,
+                datacenter=datacenter,
+                timeout=timeout,
+            )
+        return cls._instances[name]
+
+    def __init__(
+        self,
+        audience,
+        issuers,
+        timeout=None,
+    ):
+        assert audience  # ensure an audience is set always
+        self.audience = audience
+        self.issuers = issuers
+        self.keys = {issuer: None for issuer in self.issuers}
+        self._session = requests.Session()
+        self._session_adapter = TimeoutRetryingHTTPAdapter(timeout=timeout)
+        self._session.mount("http://", self._session_adapter)
+        self._session.mount("https://", self._session_adapter)
+
+    def _make_issuer_request(self, issuer):
+        headers = {
+            "X-Vault-Request": "true",  # be compatible with emissary
```

Comment:
Mention in this comment this is for avoiding SSRF attacks when making requests against Vault, or just simply "be compatible with Vault"

Reply:
done